### PR TITLE
feat: New Module `avm/res/sql/server`

### PR DESCRIPTION
**Pull request by krbar**
## Description

New Module `avm/res/sql/server`, migrated from CARML.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [ ] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [ ] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

## Pipeline references

| Pipeline |
| - |
| [avm.res.sql.server](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.sql.server.yml) |

### CODE REVIEW

File: avm/res/sql/server/database/backup-short-term-retention-policy/main.bicep [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+param retentionDays int = 7
+
+resource server 'Microsoft.Sql/servers@2022-05-01-preview' existing = {
+  name: serverName
+}
+
+resource database 'Microsoft.Sql/servers/databases@2022-05-01-preview' existing = {
+  name: databaseName
+  parent: server
+}
```
Comment (AlexanderSehr) [inline avm/res/sql/server/database/backup-short-term-retention-policy/main.bicep:24]:
```suggestion
resource server 'Microsoft.Sql/servers@2022-05-01-preview' existing = {
  name: serverName

  resource database 'databases@2022-05-01-preview' existing = {
    name: databaseName
  }
}
```

File: avm/res/sql/server/database/backup-short-term-retention-policy/main.bicep [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+resource backupShortTermRetentionPolicy 'Microsoft.Sql/servers/databases/backupShortTermRetentionPolicies@2022-05-01-preview' = {
+  name: 'default'
+  parent: database
```
Comment (AlexanderSehr) [inline avm/res/sql/server/database/backup-short-term-retention-policy/main.bicep:28]:
```suggestion
  parent: server::database
```

File: avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+param yearlyRetention string = ''
+
+resource server 'Microsoft.Sql/servers@2022-05-01-preview' existing = {
+  name: serverName
+}
+
+resource database 'Microsoft.Sql/servers/databases@2022-05-01-preview' existing = {
+  name: databaseName
+  parent: server
+}
```
Comment (AlexanderSehr) [inline avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep:30]:
```suggestion
resource server 'Microsoft.Sql/servers@2022-05-01-preview' existing = {
  name: serverName

  resource database 'databases@2022-05-01-preview' existing = {
    name: databaseName
  }
}
```

File: avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+resource backupLongTermRetentionPolicy 'Microsoft.Sql/servers/databases/backupLongTermRetentionPolicies@2022-05-01-preview' = {
+  name: 'default'
+  parent: database
```
Comment (AlexanderSehr) [inline avm/res/sql/server/database/backup-long-term-retention-policy/main.bicep:34]:
```suggestion
  parent: server::database
```

File: avm/res/sql/server/database/main.bicep [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+param collation string = 'SQL_Latin1_General_CP1_CI_AS'
+
+@description('Optional. The skuTier or edition of the particular SKU.')
+param skuTier string = 'GeneralPurpose'
+
+@description('Optional. The name of the SKU.')
+param skuName string = 'GP_Gen5_2'
+
+@description('Optional. Capacity of the particular SKU.')
+param skuCapacity int = -1
```
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:21]: Please change the implamentation to use nullable. To do so, you can make this parameter nullable, update the place where it is used to not check for `-1` and also update the reference in the parent

Comment (AlexanderSehr) [reply avm/res/sql/server/database/main.bicep:21]: It's in fact one of the biggest upsides for nullable

File: avm/res/sql/server/database/main.bicep [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+@description('The resource ID of the deployed database.')
+output resourceId string = database.id
+
+@description('The resource group of the deployed database.')
+output resourceGroupName string = resourceGroup().name
+
+@description('The location the resource was deployed into.')
+output location string = database.location
+// =============== //
```
Comment (AlexanderSehr) [inline avm/res/sql/server/database/main.bicep:228]:
```suggestion

// =============== //
```

File: avm/res/sql/server/elastic-pool/main.bicep [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+param skuCapacity int = 2
+
+@description('Optional. The name of the SKU, typically, a letter + Number code, e.g. P3.')
+param skuName string = 'GP_Gen5'
+
+@description('Optional. The tier or edition of the particular SKU, e.g. Basic, Premium.')
+param skuTier string = 'GeneralPurpose'
+
+@description('Optional. The number of secondary replicas associated with the elastic pool that are used to provide high availability. Applicable only to Hyperscale elastic pools.')
+param highAvailabilityReplicaCount int = -1
```
Comment (AlexanderSehr) [inline avm/res/sql/server/elastic-pool/main.bicep:27]: Same comment as for the database module

File: avm/res/sql/server/elastic-pool/main.bicep [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+param licenseType string = 'LicenseIncluded'
+
+@description('Optional. Maintenance configuration resource ID assigned to the elastic pool. This configuration defines the period when the maintenance updates will will occur.')
+param maintenanceConfigurationId string = ''
+
+@description('Optional. The storage limit for the database elastic pool in bytes.')
+param maxSizeBytes int = 34359738368
+
+@description('Optional. Minimal capacity that serverless pool will not shrink below, if not paused.')
+param minCapacity int = -1
```
Comment (AlexanderSehr) [inline avm/res/sql/server/elastic-pool/main.bicep:43]: Same comment as for the database module

File: avm/res/sql/server/key/main.bicep [hunk 9]
```diff
@@ -0,0 +1,47 @@
+metadata name = 'Azure SQL Server Keys'
+metadata description = 'This module deploys an Azure SQL Server Key.'
+metadata owner = 'Azure/module-maintainers'
+
+@description('Required. The name of the key. Must follow the [<keyVaultName>_<keyName>_<keyVersion>] pattern.')
+param name string
```
Comment (AlexanderSehr) [inline avm/res/sql/server/key/main.bicep:6]: This should either be optional or required, not both. As you can see in line 32, we actually have a default. Either we remove that, or we change the behavior of this parameter

File: avm/res/sql/server/main.bicep [hunk 10]
```diff
@@ -0,0 +1,460 @@
...
+  properties: {
+    administratorLogin: !empty(administratorLogin) ? administratorLogin : null
+    administratorLoginPassword: !empty(administratorLoginPassword) ? administratorLoginPassword : null
+    administrators: !empty(administrators) ? {
+      administratorType: 'ActiveDirectory'
+      azureADOnlyAuthentication: administrators.azureADOnlyAuthentication
+      login: administrators.login
+      principalType: administrators.principalType
+      sid: administrators.sid
+      tenantId: administrators.tenantId
```
Comment (AlexanderSehr) [inline avm/res/sql/server/main.bicep:144]:
```suggestion
      tenantId: administrators?.tenantId ?? tenant().tenantId
```

Makes it optional and enables us to remove the parameter from the `admin` test case

File: avm/res/sql/server/tests/e2e/defaults/main.test.bicep [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+// ============== //
+// Test Execution //
+// ============== //
+
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    // You parameters go here
```
Comment (AlexanderSehr) [inline avm/res/sql/server/tests/e2e/defaults/main.test.bicep:49]:
```suggestion
```

File: avm/res/sql/server/vulnerability-assessment/main.bicep [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+param createStorageRoleAssignment bool = true
+
+resource server 'Microsoft.Sql/servers@2022-05-01-preview' existing = {
+  name: serverName
+}
+
+// Assign SQL Server MSI access to storage account
+module storageAccount_sbdc_rbac 'modules/nested_storageRoleAssignment.bicep' = if (!useStorageAccountAccessKey && createStorageRoleAssignment) {
+  name: '${server.name}-sbdc-rbac'
+  scope: resourceGroup(split(storageAccountResourceId, '/')[4])
```
Comment (AlexanderSehr) [inline avm/res/sql/server/vulnerability-assessment/main.bicep:36]: Unless there is a rule that the SA must be in the same subscription, we should probably change this to

```suggestion
  scope: resourceGroup(split(storageAccountResourceId, '/')[2], split(storageAccountResourceId, '/')[4])
```

File: avm/res/sql/server/elastic-pool/main.bicep (hunk 7 above)
Comment (krbar) [reply avm/res/sql/server/elastic-pool/main.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/637/commits/7f39c92505cf1fe46baddb6130f9d0916fa2efea

File: avm/res/sql/server/database/main.bicep (hunk 5 above)
Comment (krbar) [reply avm/res/sql/server/database/main.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/637/commits/373c68d9d04b531aa3673ea250d4b94c035c4dca

Comment (krbar) [reply avm/res/sql/server/database/main.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/637/commits/373c68d9d04b531aa3673ea250d4b94c035c4dca and https://github.com/Azure/bicep-registry-modules/pull/637/commits/116cd0c7ca945bcc8d0714bfe432a7833f46f305

File: avm/res/sql/server/key/main.bicep (hunk 9 above)
Comment (krbar) [reply avm/res/sql/server/key/main.bicep]: Done: I made the name parameter optional (nullable) and converted one of the tests to deploy keys with a default name. See: https://github.com/Azure/bicep-registry-modules/pull/637/commits/4c53d6ce1d4db91ca24958e5311e7b09c43ed128 and https://github.com/Azure/bicep-registry-modules/pull/637/commits/3241ddb6769846178ace5ab1485edd344d832ca6

File: avm/res/sql/server/main.bicep (hunk 10 above)
Comment (krbar) [reply avm/res/sql/server/main.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/637/commits/e1b0a14687daf89f9bcc72b4b46c3a5aa67956ff

File: avm/res/sql/server/elastic-pool/main.bicep (hunk 8 above)
Comment (krbar) [reply avm/res/sql/server/elastic-pool/main.bicep]: done: https://github.com/Azure/bicep-registry-modules/pull/637/commits/8ca7dc449f84bcee7942c205cf4daf1ebb88a752

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**